Build failure: ceph

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

- Build *ceph*

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/296952995

### Relevant log output

```console
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h: In member function 'virtual std::unique_ptr<parquet::PageReader> parquet::ceph::SerializedRowGroup::GetColumnPageReader(int)':
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h:995:50: error: could not convert 'meta_decryptor' from 'std::shared_ptr<parquet::Decryptor>' to 'std::function<std::unique_ptr<parquet::Decryptor>()>'
  995 |                         static_cast<int16_t>(i), meta_decryptor, data_decryptor);
      |                                                  ^~~~~~~~~~~~~~
      |                                                  |
      |                                                  std::shared_ptr<parquet::Decryptor>
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h:1013:48: error: could not convert 'meta_decryptor' from 'std::shared_ptr<parquet::Decryptor>' to 'std::function<std::unique_ptr<parquet::Decryptor>()>'
 1013 |                       static_cast<int16_t>(i), meta_decryptor, data_decryptor);
      |                                                ^~~~~~~~~~~~~~
      |                                                |
      |                                                std::shared_ptr<parquet::Decryptor>
```
```console
error: builder for '/nix/store/bc8b5vik9l7ajkph1m85x5wnrcnyqc1n-ceph-19.2.2.drv' failed with exit code 2;
       last 25 log lines:
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > In file included from /build/ceph-19.2.2/src/log/Log.h:19,
       >                  from /build/ceph-19.2.2/src/common/dout.h:33:
       > /build/ceph-19.2.2/src/log/Entry.h: In member function 'void librbd::crypto::luks::LoadRequest<I>::finish(int) [with I = librbd::ImageCtx]':
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       >    35 |     strncpy(m_thread_name, Thread::get_thread_name().data(), 16);
       >       |            ^
       > /build/ceph-19.2.2/src/log/Entry.h: In member function 'bool librbd::crypto::luks::LoadRequest<I>::handle_read(int) [with I = librbd::ImageCtx]':
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h: In member function 'void librbd::crypto::luks::LoadRequest<I>::handle_read_header(int) [with I = librbd::ImageCtx]':
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h: In member function 'void librbd::crypto::luks::LoadRequest<I>::handle_read_keyslots(int) [with I = librbd::ImageCtx]':
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > /build/ceph-19.2.2/src/log/Entry.h: In member function 'void librbd::util::detail::C_CallbackAdapter<T, MF>::finish(int) [with T = librbd::crypto::luks::LoadRequest<librbd::ImageCtx>; void (T::* MF)(int) = &librbd::crypto::luks::LoadRequest<librbd::ImageCtx>::handle_read_keyslots]':
       > /build/ceph-19.2.2/src/log/Entry.h:35:12: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound 16 equals destination size []
       > [ 84%] Linking CXX static library ../../lib/librbd_internal.a
       > [ 84%] Built target rbd_internal
       > make: *** [Makefile:146: all] Error 2
       For full logs, run 'nix log /nix/store/bc8b5vik9l7ajkph1m85x5wnrcnyqc1n-ceph-19.2.2.drv'.
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.13.12-xanmod1, NixOS, 24.11 (Vicuna), 24.11.20250506.1d3aeb5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.14`
 - nixpkgs: `/nix/store/lzml2qh553njgdalw5cbbc208vwachr6-source`

### Notify maintainers

@adevress 
@alexanderkjeldaas 
@johanot 
@krav 
@nh2 
@benaryorg 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

I can confirm that this happens, already running a branch against *master* in anticipation of 25.05, [my own Hydra reported a failure too](https://hydra.cloud.bsocat.net/build/307428/nixlog/49).
There are a few changes in the [respective diff causing the failure](https://hydra.cloud.bsocat.net/api/scmdiff?uri=https%3A%2F%2Fgit.shell.bsocat.net%2Fnixpkgs&branch=&type=git&rev2=d89fc19e405cb2d55ce7cc114356846a0ee5e956&rev1=dda3dcd3fe03e991015e9a74b22d35950f264a54) but given the below error I think e59c1b4f9d7fdd6f01b22f4eca9d504e9eb2998f is the most promising one as [the ceph derivation includes arrow](https://github.com/NixOS/nixpkgs/blob/56ad754f5ba91a6f06fce0553fe3b2be7a178204/pkgs/tools/filesystems/ceph/default.nix#L28).

<details><summary>first compile error</summary>

```text
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h: In member function 'virtual std::unique_ptr<parquet::PageReader> parquet::ceph::SerializedRowGroup::GetColumnPageReader(int)':
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h:995:50: error: could not convert 'meta_decryptor' from 'std::shared_ptr<parquet::Decryptor>' to 'std::function<std::unique_ptr<parquet::Decryptor>()>'
  995 |                         static_cast<int16_t>(i), meta_decryptor, data_decryptor);
      |                                                  ^~~~~~~~~~~~~~
      |                                                  |
      |                                                  std::shared_ptr<parquet::Decryptor>
/build/ceph-19.2.2/src/s3select/include/s3select_parquet_intrf.h:1013:48: error: could not convert 'meta_decryptor' from 'std::shared_ptr<parquet::Decryptor>' to 'std::function<std::unique_ptr<parquet::Decryptor>()>'
 1013 |                       static_cast<int16_t>(i), meta_decryptor, data_decryptor);
      |                                                ^~~~~~~~~~~~~~
      |                                                |
      |                                                std::shared_ptr<parquet::Decryptor>
```

</details>

I could run a bisect (given my local node has seen a CPU and memory upgrade the day before yesterday, what timing) in case this isn't the commit, but given the power consumption I'd rather wait until that one commit is verified to not be the issue ^^

For now I'll leave it at that, but I'll check back on this tomorrow (though I have a feeling @nh2 will beat me to it).

--- a/pkgs/tools/filesystems/ceph/arrow-cpp-19.nix
+++ b/pkgs/tools/filesystems/ceph/arrow-cpp-19.nix
@@ -1,0 +1,335 @@
+# This is is arrow-cpp < 20 used as a workaround for
+# Ceph not supporting >= yet, taken from nixpkgs commit
+#     97ae53798f6a7c7c3c259ad8c2cbcede6ca34b2a~
+# This should be entirely removed when upstream bug
+#     https://tracker.ceph.com/issues/71269
+# is fixed.
+{
+  stdenv,
+  lib,
+  fetchurl,
+  fetchFromGitHub,
+  fixDarwinDylibNames,
+  autoconf,
+  aws-sdk-cpp,
+  aws-sdk-cpp-arrow ? aws-sdk-cpp.override {
+    apis = [
+      "cognito-identity"
+      "config"
+      "identity-management"
+      "s3"
+      "sts"
+      "transfer"
+    ];
+  },
+  boost,
+  brotli,
+  bzip2,
+  cmake,
+  crc32c,
+  curl,
+  flatbuffers,
+  gflags,
+  glog,
+  google-cloud-cpp,
+  grpc,
+  gtest,
+  libbacktrace,
+  lz4,
+  minio,
+  ninja,
+  nlohmann_json,
+  openssl,
+  perl,
+  pkg-config,
+  protobuf_29,
+  python3,
+  rapidjson,
+  re2,
+  snappy,
+  sqlite,
+  thrift,
+  tzdata,
+  utf8proc,
+  which,
+  zlib,
+  zstd,
+  testers,
+  enableShared ? !stdenv.hostPlatform.isStatic,
+  enableFlight ? stdenv.buildPlatform == stdenv.hostPlatform,
+  # Disable also on RiscV
+  # configure: error: cannot determine number of significant virtual address bits
+  enableJemalloc ?
+    !stdenv.hostPlatform.isDarwin && !stdenv.hostPlatform.isAarch64 && !stdenv.hostPlatform.isRiscV64,
+  enableS3 ? true,
+  # google-cloud-cpp fails to build on RiscV
+  enableGcs ? !stdenv.hostPlatform.isDarwin && !stdenv.hostPlatform.isRiscV64,
+}:
+
+let
+  # https://github.com/apache/arrow/issues/45807
+  protobuf = protobuf_29;
+
+  arrow-testing = fetchFromGitHub {
+    name = "arrow-testing";
+    owner = "apache";
+    repo = "arrow-testing";
+    rev = "4d209492d514c2d3cb2d392681b9aa00e6d8da1c";
+    hash = "sha256-IkiCbuy0bWyClPZ4ZEdkEP7jFYLhM7RCuNLd6Lazd4o=";
+  };
+
+  parquet-testing = fetchFromGitHub {
+    name = "parquet-testing";
+    owner = "apache";
+    repo = "parquet-testing";
+    rev = "c7cf1374cf284c0c73024cd1437becea75558bf8";
+    hash = "sha256-DThjyZ34LajHwXZy1IhYKUGUG/ejQ9WvBNuI8eUKmSs=";
+  };
+
+  version = "19.0.1";
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "arrow-cpp";
+  inherit version;
+
+  src = fetchFromGitHub {
+    owner = "apache";
+    repo = "arrow";
+    rev = "apache-arrow-${version}";
+    hash = "sha256-toHwUIOZRpgR0K7pQtT5nqWpO9G7AuHYTcvA6UVg9lA=";
+  };
+
+  sourceRoot = "${finalAttrs.src.name}/cpp";
+
+  # versions are all taken from
+  # https://github.com/apache/arrow/blob/apache-arrow-${version}/cpp/thirdparty/versions.txt
+
+  # jemalloc: arrow uses a custom prefix to prevent default allocator symbol
+  # collisions as well as custom build flags
+  ${if enableJemalloc then "ARROW_JEMALLOC_URL" else null} = fetchurl {
+    url = "https://github.com/jemalloc/jemalloc/releases/download/5.3.0/jemalloc-5.3.0.tar.bz2";
+    hash = "sha256-LbgtHnEZ3z5xt2QCGbbf6EeJvAU3mDw7esT3GJrs/qo=";
+  };
+
+  # mimalloc: arrow uses custom build flags for mimalloc
+  ARROW_MIMALLOC_URL = fetchFromGitHub {
+    owner = "microsoft";
+    repo = "mimalloc";
+    rev = "v2.0.6";
+    hash = "sha256-u2ITXABBN/dwU+mCIbL3tN1f4c17aBuSdNTV+Adtohc=";
+  };
+
+  ARROW_XSIMD_URL = fetchFromGitHub {
+    owner = "xtensor-stack";
+    repo = "xsimd";
+    rev = "13.0.0";
+    hash = "sha256-qElJYW5QDj3s59L3NgZj5zkhnUMzIP2mBa1sPks3/CE=";
+  };
+
+  ARROW_SUBSTRAIT_URL = fetchFromGitHub {
+    owner = "substrait-io";
+    repo = "substrait";
+    rev = "v0.44.0";
+    hash = "sha256-V739IFTGPtbGPlxcOi8sAaYSDhNUEpITvN9IqdPReug=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    ninja
+    autoconf # for vendored jemalloc
+    flatbuffers
+  ] ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames;
+  buildInputs =
+    [
+      boost
+      brotli
+      bzip2
+      flatbuffers
+      gflags
+      glog
+      gtest
+      libbacktrace
+      lz4
+      nlohmann_json # alternative JSON parser to rapidjson
+      protobuf # substrait requires protobuf
+      rapidjson
+      re2
+      snappy
+      thrift
+      utf8proc
+      zlib
+      zstd
+    ]
+    ++ lib.optionals enableFlight [
+      grpc
+      openssl
+      protobuf
+      sqlite
+    ]
+    ++ lib.optionals enableS3 [
+      aws-sdk-cpp-arrow
+      openssl
+    ]
+    ++ lib.optionals enableGcs [
+      crc32c
+      curl
+      google-cloud-cpp
+      grpc
+      nlohmann_json
+    ];
+
+  preConfigure = ''
+    patchShebangs build-support/
+    substituteInPlace "src/arrow/vendored/datetime/tz.cpp" \
+      --replace-fail 'discover_tz_dir();' '"${tzdata}/share/zoneinfo";'
+  '';
+
+  cmakeFlags =
+    [
+      "-DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON"
+      "-DARROW_BUILD_SHARED=${if enableShared then "ON" else "OFF"}"
+      "-DARROW_BUILD_STATIC=${if enableShared then "OFF" else "ON"}"
+      "-DARROW_BUILD_TESTS=${if enableShared then "ON" else "OFF"}"
+      "-DARROW_BUILD_INTEGRATION=ON"
+      "-DARROW_BUILD_UTILITIES=ON"
+      "-DARROW_EXTRA_ERROR_CONTEXT=ON"
+      "-DARROW_VERBOSE_THIRDPARTY_BUILD=ON"
+      "-DARROW_DEPENDENCY_SOURCE=SYSTEM"
+      "-Dxsimd_SOURCE=AUTO"
+      "-DARROW_DEPENDENCY_USE_SHARED=${if enableShared then "ON" else "OFF"}"
+      "-DARROW_COMPUTE=ON"
+      "-DARROW_CSV=ON"
+      "-DARROW_DATASET=ON"
+      "-DARROW_FILESYSTEM=ON"
+      "-DARROW_FLIGHT_SQL=${if enableFlight then "ON" else "OFF"}"
+      "-DARROW_HDFS=ON"
+      "-DARROW_IPC=ON"
+      "-DARROW_JEMALLOC=${if enableJemalloc then "ON" else "OFF"}"
+      "-DARROW_JSON=ON"
+      "-DARROW_USE_GLOG=ON"
+      "-DARROW_WITH_BACKTRACE=ON"
+      "-DARROW_WITH_BROTLI=ON"
+      "-DARROW_WITH_BZ2=ON"
+      "-DARROW_WITH_LZ4=ON"
+      "-DARROW_WITH_NLOHMANN_JSON=ON"
+      "-DARROW_WITH_SNAPPY=ON"
+      "-DARROW_WITH_UTF8PROC=ON"
+      "-DARROW_WITH_ZLIB=ON"
+      "-DARROW_WITH_ZSTD=ON"
+      "-DARROW_MIMALLOC=ON"
+      "-DARROW_SUBSTRAIT=ON"
+      "-DARROW_FLIGHT=${if enableFlight then "ON" else "OFF"}"
+      "-DARROW_FLIGHT_TESTING=${if enableFlight then "ON" else "OFF"}"
+      "-DARROW_S3=${if enableS3 then "ON" else "OFF"}"
+      "-DARROW_GCS=${if enableGcs then "ON" else "OFF"}"
+      # Parquet options:
+      "-DARROW_PARQUET=ON"
+      "-DPARQUET_BUILD_EXECUTABLES=ON"
+      "-DPARQUET_REQUIRE_ENCRYPTION=ON"
+    ]
+    ++ lib.optionals (!enableShared) [ "-DARROW_TEST_LINKAGE=static" ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      "-DCMAKE_INSTALL_RPATH=@loader_path/../lib" # needed for tools executables
+    ]
+    ++ lib.optionals (!stdenv.hostPlatform.isx86_64) [ "-DARROW_USE_SIMD=OFF" ]
+    ++ lib.optionals enableS3 [
+      "-DAWSSDK_CORE_HEADER_FILE=${aws-sdk-cpp-arrow}/include/aws/core/Aws.h"
+    ];
+
+  doInstallCheck = true;
+  ARROW_TEST_DATA = lib.optionalString finalAttrs.doInstallCheck "${arrow-testing}/data";
+  PARQUET_TEST_DATA = lib.optionalString finalAttrs.doInstallCheck "${parquet-testing}/data";
+  GTEST_FILTER =
+    let
+      # Upstream Issue: https://issues.apache.org/jira/browse/ARROW-11398
+      filteredTests =
+        lib.optionals stdenv.hostPlatform.isAarch64 [
+          "TestFilterKernelWithNumeric/3.CompareArrayAndFilterRandomNumeric"
+          "TestFilterKernelWithNumeric/7.CompareArrayAndFilterRandomNumeric"
+          "TestCompareKernel.PrimitiveRandomTests"
+        ]
+        ++ lib.optionals enableS3 [
+          "S3OptionsTest.FromUri"
+          "S3RegionResolutionTest.NonExistentBucket"
+          "S3RegionResolutionTest.PublicBucket"
+          "S3RegionResolutionTest.RestrictedBucket"
+          "TestMinioServer.Connect"
+          "TestS3FS.*"
+          "TestS3FSGeneric.*"
+        ]
+        ++ lib.optionals stdenv.hostPlatform.isDarwin [
+          # TODO: revisit at 12.0.0 or when
+          # https://github.com/apache/arrow/commit/295c6644ca6b67c95a662410b2c7faea0920c989
+          # is available, see
+          # https://github.com/apache/arrow/pull/15288#discussion_r1071244661
+          "ExecPlanExecution.StressSourceSinkStopped"
+        ];
+    in
+    lib.optionalString finalAttrs.doInstallCheck "-${lib.concatStringsSep ":" filteredTests}";
+
+  __darwinAllowLocalNetworking = true;
+
+  nativeInstallCheckInputs =
+    [
+      perl
+      which
+      sqlite
+    ]
+    ++ lib.optionals enableS3 [ minio ]
+    ++ lib.optionals enableFlight [ python3 ];
+
+  installCheckPhase =
+    let
+      disabledTests = [
+        # flaky
+        "arrow-flight-test"
+        # requires networking
+        "arrow-gcsfs-test"
+        "arrow-flight-integration-test"
+      ];
+    in
+    ''
+      runHook preInstallCheck
+
+      ctest -L unittest --exclude-regex '^(${lib.concatStringsSep "|" disabledTests})$'
+
+      runHook postInstallCheck
+    '';
+
+  meta = with lib; {
+    description = "Cross-language development platform for in-memory data";
+    homepage = "https://arrow.apache.org/docs/cpp/";
+    license = licenses.asl20;
+    platforms = platforms.unix;
+    maintainers = with maintainers; [
+      tobim
+      veprbl
+      cpcloud
+    ];
+    pkgConfigModules = [
+      "arrow"
+      "arrow-acero"
+      "arrow-compute"
+      "arrow-csv"
+      "arrow-dataset"
+      "arrow-filesystem"
+      "arrow-flight"
+      "arrow-flight-sql"
+      "arrow-flight-testing"
+      "arrow-json"
+      "arrow-substrait"
+      "arrow-testing"
+      "parquet"
+    ];
+  };
+  passthru = {
+    inherit
+      enableFlight
+      enableJemalloc
+      enableS3
+      enableGcs
+      ;
+    tests.pkg-config = testers.testMetaPkgConfig finalAttrs.finalPackage;
+  };
+})

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2548,6 +2548,11 @@ with pkgs;
       # If we want to switch for that before upstream fixes it, use this patch:
       # https://github.com/NixOS/nixpkgs/pull/281858#issuecomment-1899648638
       fmt = fmt_9;
+
+      # Remove once Ceph supports arrow-cpp >= 20, see:
+      # * https://tracker.ceph.com/issues/71269
+      # * https://github.com/NixOS/nixpkgs/issues/406306
+      arrow-cpp = callPackage ../tools/filesystems/ceph/arrow-cpp-19.nix { };
     })
     ceph
     ceph-client

